### DISCUSSION TITLE
Signal url couldnt be an expression

### DISCUSSION THREAD

**Discussion by rezabashiri** [elsa-workflows/elsa-core#1289] [Q&A]
At 'Signal' activity, elsa just get variable and compare it by the signal property, but it doesn't respect type signal, it could be java expression so I suggest an easy correction tp overcome this kind of lacking 

```csharp
protected override async Task<bool> OnCanExecuteAsync(WorkflowExecutionContext context, CancellationToken cancellationToken)
{
    var signal = await expressionEvaluator.EvaluateAsync(Signal, context, cancellationToken);
    var expression = new JavaScriptExpression<string>($"`{context.Workflow.Input.GetVariable<string>("Signal")}`");
    var signaled =await expressionEvaluator.EvaluateAsync<string>(expression,context, cancellationToken);
    return signaled == signal;
}
```

**Comment by craigfowler** [comment]
Hi @rezabashiri I'm sorry but I'm not really sure I understand your question, or maybe you trying to suggest a feature.
Could you give a little bit more information please?

(I edited your code sample - on Github you can use three back-ticks: ``` to mark a code-block)

**Comment by sfmskywalker** [comment]
I think I do understand actually. It looks like @rezabashiri wants to evaluate the **signal** provided as **workflow input** as well:

`var expression = new JavaScriptExpression<string>($"``{context.Workflow.Input.GetVariable<string>("Signal")}``");`

However, the suggested change will require the input to **always** be a valid **JavaScript** expression. Additionally, evaluating workflow input like this in a given activity seems too arbitrary for my sanity.

@rezabashiri Perhaps if you would share with us your use case, we can think of a better solution if we need one?

**Comment by rezabashiri** [comment]
@sfmskywalker  yeah this is some kinda enhancement, I did a small hack to support both type of literal and javascript expressions 


`
        protected override async Task<bool> OnCanExecuteAsync(WorkflowExecutionContext context, CancellationToken cancellationToken)
        {
            var signal = await expressionEvaluator.EvaluateAsync(Signal, context, cancellationToken);
            var signaled = await ExtractSignal(context, cancellationToken);
            return signaled == signal;
        }

        public async static Task<string> ExtractSignal(WorkflowExecutionContext context,    CancellationToken cancellationToken)
        {
            var received = context.Workflow.Input.GetVariable<string>("Signal");
            if (string.IsNullOrEmpty(received))
                return string.Empty;
            var syntax = received.Contains('$') switch
            {
                false => "Literal",
                true => "JavaScript"
            };
            var expression = syntax == "JavaScript" ? new JavaScriptExpression<string>($"`{received}`") : new WorkflowExpression<string>(syntax, received);
            var signaled = await context.EvaluateAsync<string>(expression, cancellationToken);
            return signaled;
        }`

**Comment by sfmskywalker** [comment]
With Elsa 2, it is now possible to define dynamically evaluated signal names. A sample of this can be found in [this guide](https://elsa-workflows.github.io/elsa-core/docs/next/guides/guides-parent-child-workflows#child-workflow-2).
